openvasd/openvasd.c

```diff
@@ -0,0 +1,2251 @@
+{
+  s->len = 0;
+  s->ptr = g_malloc0 (s->len + 1);
+  if (s->ptr == NULL)
+    {
+      g_warning ("%s: Error allocating memory for response", __func__);
+      return;
+    }
+  s->ptr[0] = '\0';
+}
+
+/** @brief Reinitialize the string struct to hold the response
+ *
+ *  @param s[in/out] The string struct to be initialized
+ */
+static void
+reset_openvasd_stringstream (openvasd_stringstream *s)
+{
+  g_free (s->ptr);
+  init_openvasd_stringstream (s);
+}
+
+/** @brief Call back function to stored the response.
+ *
+ *  @description The function signature is the necessary to work with
+ *  libcurl. It stores the response in s. It reallocate memory if necessary.
+ */
+static size_t
+response_callback_fn (void *ptr, size_t size, size_t nmemb, void *struct_string)
+{
+  openvasd_stringstream *s = struct_string;
+  size_t new_len = s->len + size * nmemb;
+  char *ptr_aux = g_realloc (s->ptr, new_len + 1);
+  s->ptr = ptr_aux;
+  if (s->ptr == NULL)
+    {
+      g_warning ("%s: Error allocating memory for response", __func__);
+      return 0; // no memory left
+    }
+  memcpy (s->ptr + s->len, ptr, size * nmemb);
+  s->ptr[new_len] = '\0';
+  s->len = new_len;
+
+  return size * nmemb;
+}
+
+/** @brief Create a CURL handler
+ *
+ *  @param[in] conn    struct holding the openvasd connector information
+ *  @param[in] method  request method (e.g. GET)
+ *  @param[in] path    Path to the resource (e.g. /vts)
+ *  @param[in] data    String containing the request body in json format (scan
+ * action, scan config)
+ *  @param[out] resp   Structure holding the body response, filled by the
+ * callback function
+ *  @param[out] err    On error, this variable is filled with an error message
+ * in json format.
+ *
+ * @return a CURL handler, or NULL on error.
+ */
+static CURL *
+handler (openvasd_connector_t *conn, openvasd_req_method_t method, char *path,
+         char *data, openvasd_stringstream *resp, char **err)
+{
+  CURL *curl;
+  GString *url = NULL;
+  struct curl_slist *customheader = NULL;
+  GString *xapikey = NULL;
+
+  if (!conn)
+    {
+      *err = g_strdup ("{\"error\": \"Missing openvasd connector\"}");
+      g_warning ("%s: Missing openvasd connector", __func__);
+      return NULL;
+    }
+
+  if ((curl = curl_easy_init ()) == NULL)
+    {
+      *err =
+        g_strdup ("{\"error\": \"Not possible to initialize curl library\"}");
+      g_warning ("%s: Not possible to initialize curl library", __func__);
+      return NULL;
+    }
+
+  url = g_string_new (g_strdup ((*conn)->server));
+
+  if ((*conn)->port > 0 && (*conn)->port < 65535)
+    {
+      char buf[6];
+      g_snprintf (buf, sizeof (buf), ":%d", (*conn)->port);
+      g_string_append (url, buf);
+    }
+
+  if (path != NULL && path[0] != '\0')
+    g_string_append (url, path);
+
+  // Set URL
+  g_debug ("%s: URL: %s", __func__, url->str);
+  if (curl_easy_setopt (curl, CURLOPT_URL, g_strdup (url->str)) != CURLE_OK)
```

Comment:
`curl_easy_setopt` says

    Strings passed to libcurl as 'char *' arguments, are copied by the library

so this is leaking that g_strdup.

Also url should be freed on failure.